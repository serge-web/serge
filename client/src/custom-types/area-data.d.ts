import ForceData from './force-data'

/** a user-defined area, used for organising activities */
export interface Area {
  // user name for this area
  name: string
  // the bounds of the area. Note that GeoJSON stores them in lng/lat order
  polygon: GeoJSON.Polygon
}

export interface AreaCategory {
  // name for this category
  name: string
  // color for this category
  color: string
  // areas
  areas: Array<Area>
  // which forces use this category
  usedBy: Array<ForceData['uniqid']>
}

/** the parent wrapper for the set of areas */
export default interface AreaData {
  areas: Array<AreaCategory>
}
