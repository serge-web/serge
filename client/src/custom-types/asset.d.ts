import { LaydownTypes } from 'src/config'
import { AttributeValues, AttributeValues2, Perception, PlatformTypeData, Role, RouteStatus, RouteTurn } from '.'

export interface Asset {
  /**
   * the name of this asset
   */
  name: string
  /** when an asset is first identified, only the
   * contact id is shown.  Something like `C023`.
   */
  readonly contactId: string
  /** the uniq id of this asset, present to allow
   * assets to be renamed, or multiple assets of same
   * name present in game
   */
  readonly uniqid: string
  /** the owner of this asset. This is used to give 
   * control of the asset to a specific player
   */
  owner?: Role['roleId']
  /** the current position of an asset. It's optional,
   * since an asset may be carried by another
   */
  position?: string
  /** the geographic position of an asset.
   * If it's not present, but a `position` is, then
   * the location will be generated. Sequence is `Lat, Long`
   */
  location?: [number, number]
  /** how other forces perceive this asset
   */
  perceptions: Array<Perception>
  /** where this platform has already been
   */
  history?: Array<RouteTurn>
  /** the set of planned turns for this asset
   */
  plannedTurns?: Array<RouteTurn>
  /** the current condition of this asset
   */
  condition: string
  /** 
   * the type-id of this platform 
   */
  platformTypeId: PlatformTypeData['uniqid']
  /** if the player is allowed to change the position of this asset
   * during turn-zero of a wargame
   * note: we allow string so it can be set in JSON
   */
  locationPending?: LaydownTypes | 'Force laydown' | 'Umpire laydown' 
  /** whether this asset has been destroyoed
   * //TODO: move away from this, use condition
   */
  destroyed?: boolean
  /** if this asset is actually an organisational group
   * of assets
   */
  comprising?: Array<Asset>
  /** if this asset is carrying other assets
   */
  hosting?: Array<Asset>
  /** current status of this asset
   *
  */
  status?: RouteStatus
  /** attributes for this asset 
   */
  attributeValues?: AttributeValues
  /** attributes for this asset, using the streamlined notation
   */
  attributes?: AttributeValues2
   /** health of this asset. 0-100, with 0 treated as destroyed */
  health?: number
}

export default Asset
