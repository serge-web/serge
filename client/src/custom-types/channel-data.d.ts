import { CHANNEL_CHAT, CHANNEL_COLLAB, CHANNEL_CUSTOM, CHANNEL_MAPPING, InitialStates, SpecialChannelColumns } from 'src/config'
import MappingConstraints from './mapping-constraints'
import { ParticipantChat, ParticipantCollab, ParticipantCustom, ParticipantMapping, ParticipantTemplate } from './participant'

/** description of channel, as stored in database */
export interface ChannelCore {
  /** unique id for this channel */
  readonly uniqid: string
  /** name of this channel */
  name: string
  /** type - extended by child implementations */
  readonly channelType: any
}

/** description of chat channel */
export interface ChannelChat extends ChannelCore {
  readonly channelType: typeof CHANNEL_CHAT
  /** list of participants for this channel */
  participants: Array<ParticipantChat>
  /** whether to hide the Force and Role of the message author */
  hideMessageAuthor?: boolean
}

export interface ChannelMapping extends ChannelCore {
  readonly channelType: typeof CHANNEL_MAPPING
  /** map/tile constraints */
  constraints: MappingConstraints
  /** list of participants for this channel */
  participants: Array<ParticipantMapping>
  /** list of renderers for this channel */
  renderers: Array<BaseRenderer>
}

/** description of standard channel, sending custom messages */
export interface ChannelCustom extends ChannelCore {
  readonly channelType: typeof CHANNEL_CUSTOM
  /** list of participants for this channel */
  participants: Array<ParticipantCustom>
}

/** description of collaborative editing channel */
export interface ChannelCollab extends ChannelCore {
  readonly channelType: typeof CHANNEL_COLLAB
  /** list of participants for this channel */
  participants: Array<ParticipantCollab>
  /** verbs for returning for edit */
  requestChangesVerbs: Array<string>
  /** verbs for approving document */
  approveVerbs: Array<string>
  /** verbs for releasing document */
  releaseVerbs: Array<string>
  /** who original document goes to first */
  initialState: InitialStates
  /** extra columns to show in status view */
  extraColumns: Array<SpecialChannelColumns>
  /** type for new document */
  newMessageTemplate: ParticipantTemplate | undefined
  /** type for response document */
  responseTemplate?: ParticipantTemplate
  /** additional columns to display in table.  Currently takes some 'special values',
   * see AdditionalData in genMessageCollabEdit.tsx
   */
  additonalData?: string[]
}

/** union of the assorted channel types */
export type ChannelTypes = ChannelChat |
  ChannelCustom |
  ChannelCollab |
  ChannelMapping
