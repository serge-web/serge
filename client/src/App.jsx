import { Component } from 'react'
import { Provider } from 'react-redux'
import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles'
import theme from './themes/serge'
// import Notifications from './Components/Notifications'
// import Version from './Views/Version'
import RouterMain from './Components/Router/RouterMain'
import Store from './Store/Store'
// import '@serge/themes/App.scss'
import './themes/App.scss'


class App extends Component {
  componentDidMount () {
    document.title = 'Serge'
  }
  
  render () {
    return (
      <Provider store={Store}>
        <ThemeProvider theme={theme}>
          <RouterMain />
          {/* <Notifications /> */}
          {/* <Version /> */}
        </ThemeProvider>
      </Provider>
    )
  }
}

export default App
