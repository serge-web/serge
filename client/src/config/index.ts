/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export * from './globals'
export * from './routes'
export * from './enums'
export * from './action-constants'
