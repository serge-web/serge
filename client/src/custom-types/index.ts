/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export * from './channel-data.d'
export type { default as ChannelUI } from './channel-ui.d'
export type { default as Channel } from './channel.d'
export * from './export-items-actions.d'
export * from './export-items.d'
export type { default as ForceData } from './force-data.d'
export type { default as ForceOption } from './force-option.d'
export type { default as ForceRole } from './force-role.d'
export type { default as Force } from './force.d'
export * from './global.d'
// export { default as MapPostBack } from './map-postback-data.d'
export type { default as MarkerPerception } from './marker-perception.d'
export type { default as MessageTypes, TemplateBodysByKey } from './message-types.d'
export * from './message.d'
export type { SetCurrentViewFromURI } from './messages-action.d'
export * from './participant.d'
export * from './player-log-actions.d'
export * from './player-log.d'
export * from './player-ui-actions.d'
export type { default as PlayerUi, PlayerUiChannels, PlayerUiChatChannel } from './player-ui.d'
export type { default as WargamesState } from './WargamesState.d'
export type { default as Role } from './role.d'
export * from './root-state'
export type { default as SetWargameMessage } from './set-wargame-message.d'
export type { default as Template, TemplateBody, TemplateData } from './template.d'
export * from './turn-length.d'
export * from './wargame-action.d'
export type { default as WargameData } from './wargame-data.d'
export type { default as WargameList } from './wargame-list.d'
export type { default as WargameOverview } from './wargame-overview.d'
export type { default as Wargame } from './wargame.d'
export * from './core-mapping.d'
export * from './turn-period.d'
