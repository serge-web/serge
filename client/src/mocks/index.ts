/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export { default as GameMessagesMock } from './game-messages.mock'
export { default as GameMessagesMockRFI } from './game-messages-rfi.mock'
export { default as AdminMessagesMock } from './admin-messages.mock'
export { default as ChatMessagesMock } from './chat-messages.mock'
export { default as InfoMessagesMock } from './info-messages.mock'
export { default as GameChannels2 } from './game-channels-2.mock'
export { default as WargameMock } from './wargame.mock'
export { default as WargameExportedMock } from './wargame-exported.mock'
export { default as MessageTemplatesMock } from './message-templates.mock'
export * from './message-templates-by-key.mock'
export { default as forces } from './forces.mock'
export { default as atlanticForces } from './atlantic-forces.mock'
export { default as watuWargame } from './watu-wargame-mock'
export { default as preInitialisedForces } from './pre-initialised-forces.mock'
export { default as adminTabs } from './admin-tabs.mock'
export { default as feedbackMessage, feedbackMessage1, feedbackMessage2 } from './feedback-message.mock'
export { default as actionFeedbackMessages } from './action-feedback-messages.mock'
export { default as actionSetLatestFeedbackMessage } from './action-set-latest-feedback-message.mock'
export { default as actionSetLatestWargameMessage } from './action-set-latest-wargame-message.mock'
export { default as actionSetWargameMessages } from './action-set-wargame-messages.mock'
export { default as actionOpenMessage } from './action-open-message.mock'
export { default as actionCloseMessage } from './action-close-message.mock'
export * from './coa-channel-data-2.mock'
export * from './coa-messages.mock'
