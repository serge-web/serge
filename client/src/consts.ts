import uniqId from 'uniqid'
import moment from 'moment'
import ExpiredStorage from 'expired-storage'
import * as Globals from '@serge/config'
import { Wargame } from '@serge/custom-types'

/** name of the document used to store the initial wargame definition */
export const DEFAULT_SERVER = 'Nelson'
export const DEFAULT_PORT = '8080'

export const MSG_STORE = 'messages'
export const SERGE_INFO = 'serge_info'
export const CHAT_CHANNEL_ID = 'game-admin'

export const PLANNING_PHASE = 'planning'
export const ADJUDICATION_PHASE = 'adjudication'

export const ADMIN_ROUTE = '/serge/admin'
export const MESSAGE_TEMPLATE_ROUTE = '/serge/messageTemplates'
export const MESSAGE_CREATOR_BASE_ROUTE = '/serge/messageCreator'
export const CREATE_TEMPLATE_ROUTE = '/create/template'
export const EDIT_TEMPLATE_ROUTE = '/edit/template'
export const CREATE_MESSAGE_ROUTE = '/create/message'
export const EDIT_MESSAGE_ROUTE = '/edit/message'
export const GAME_SETUP_ROUTE = '/serge/gameSetup'
export const WELCOME_SCREEN_EDIT_ROUTE = '/serge/editWelcomeScreen'
export const EXPORT_ROUTE = '/serge/export'
export const EXPORT_MESSAGES_SUBROUTE = '/messages'
export const EXPORT_FORCES_SUBROUTE = '/foeces'
export const EXPORT_PRINT_SUBROUTE = '/print/:id'
export const PLAYERUI_ROUTE = '/serge/player'
export const DEMO_ROUTE = '/serge/demo'
export const STORYBOOK_ROUTE = './storybook'

export const expiredStorage = new ExpiredStorage()
export const LOCAL_STORAGE_TIMEOUT = 2592000 // one month

export const MAX_LISTENERS = 82

export const UMPIRE_FORCE = 'umpire'

export const LOCATION_PENDING = 'LocationPending' // special state where platforms can be moved at turn zero

// return states of adjudication form
export const PLAN_ACCEPTED = 'accepted'
export const PLAN_REJECTED = 'rejected'

// names of special channels
export const CHANNEL_RFI_STATUS = 'rfis'
export const CHANNEL_MAPPING = 'mapping'

// NOTE: time period to wait if server returns an error. One frequent cause of error
// during development is that the server is stopped.  We're introducing a
// throttle value to prevent the browser going into a race condition
// as it sends 1000s of requests to the server
export const ERROR_THROTTLE = 3000

// Interval for check server heartbeats
export const SERVER_PING_INTERVAL = 20000

// Nov 2019. Ian modified the server path to use the
// current URL, so we can use Heroku to provide
// review instances of the app.  In these
// review instances, we can't predict the URL, so
// were failing CORS test
export const baseUrl = () => {
  const { hostname, protocol } = window.location
  // @ts-ignore
  const host = (new URL(window.location)).searchParams.get('host')

  // NOTE: for all non-heroku deployments, we need to append the port number
  const portDetails = hostname.toLowerCase().indexOf('herokuapp') !== -1 ? '' : ':' + DEFAULT_PORT

  const res = host || `${protocol}//${hostname}` + portDetails

  return res
}

export const serverPath = (
  // @ts-ignore
  window.G_CONFIG.REACT_APP_SERVER_PATH || process.env.REACT_APP_SERVER_PATH || baseUrl() + '/'
).replace(/\/?$/, '/')

export const databasePath = `${serverPath}db/`

export const headers = {
  'Content-Type': 'application/json',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'POST',
  'Access-Control-Allow-Headers': 'Authorization, Lang'
}

export const defaultGameInfo = {
  imageUrlSet: false,
  imageUrl: '/default_img/sergeDefault.png',
  title: 'Serge',
  description: `You have arrived at the Development Centre Gaming Facility.\n
  You will use this web-based application to interact with players from other forces, together with the umpires in the White Cell.\n
  At any point during your time here you can submit insights via the Insights button at the top-right of the gaming page.\n
  These insights could relate to the current doctrine being explored, the performance of your force, or how the game is being organised / facilitated.\n
  Thanks in advance for your participation.\n
  Maj Duncan Dare, PO1 Gaming`,
  showAccessCodes: true
}

export const forceTemplate = {
  name: '',
  uniqid: '',
  overview: 'An overview written here..',
  roles: [
    {
      name: 'CO',
      roleId: `p${uniqId.time()}`,
      isGameControl: false,
      isObserver: false,
      isInsightViewer: false
    }
  ],
  iconURL: serverPath + 'default_img/forceDefault.png',
  color: '#3dd0ff',
  umpire: false,
  dirty: false
}

export const umpireForceTemplate = {
  name: 'White',
  uniqid: 'umpire',
  overview: 'Umpire force.',
  roles: [
    {
      name: 'Game Control',
      roleId: `p${uniqId.time()}`,
      isGameControl: true,
      isObserver: true,
      isInsightViewer: true
    }
  ],
  iconURL: serverPath + 'default_img/umpireDefault.png',
  color: '#FCFBEE',
  umpire: true,
  dirty: false
}

export const channelTemplate = {
  name: '',
  uniqid: '',
  participants: []
}

export const dbDefaultPlaylogSettings = {
  wargame: 'missing',
  role: 'missing',
  activityType: {
    aType: 'pending'
  },
  activityTime: '',
  _id: new Date().toISOString(), 
  isOpen: false
}

export const dbDefaultSettings: Wargame = {
  _id: Globals.wargameSettings,
  wargameTitle: '',
  name: '',
  data: {
    overview: {
      name: 'Overview - settings',
      gameDescription: '',
      // spatialRepresentation: '',
      gameTurnTime: {
        unit: 'millis',
        millis: 43200000
      },
      realtimeTurnTime: 300000,
      timeWarning: 60000,
      // turnStrategy: '',
      gameDate: moment(new Date(), moment.ISO_8601).format(),
      showAccessCodes: true,
      logPlayerActivity: true,
      dirty: false
    },
    forces: {
      name: 'Forces',
      forces: [umpireForceTemplate],
      selectedForce: umpireForceTemplate,
      dirty: false
    },
    channels: {
      name: 'Channels',
      channels: [],
      selectedChannel: '',
      dirty: false
    }
  },
  wargameList: [],
  wargameInitiated: false,
  gameTurn: 0,
  phase: ADJUDICATION_PHASE,
  adjudicationStartTime: moment(new Date(), moment.ISO_8601).format()
}

export const FLEX_LAYOUT_MODEL_DEFAULT: any = {
  global: {
    tabSetTabStripHeight: 45,
    tabEnableClose: false,
    tabEnableRenderOnDemand: false
  },
  borders: [],
  layout: {
    type: 'row' as string,
    weight: 100,
    children: [
    ]
  }
}
