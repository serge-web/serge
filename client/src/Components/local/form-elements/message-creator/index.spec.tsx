import { ChannelUI, MessageDetails } from 'src/custom-types'
import { watuWargame, MessageTemplatesMock } from 'src/mocks'
import React from 'react'
import renderer from 'react-test-renderer'
import MessageCreator from './index'

jest.mock('leaflet', () => {
  const leafletModule = jest.requireActual('leaflet')
  const mockLeafletModule = {
    ...leafletModule,
    Symbol: {
      arrowHead: jest.fn()
    }
  }
  return mockLeafletModule
})

jest.mock('react-leaflet-v4', () => ({
  useMap: (): jest.Mock => jest.fn()
}))

jest.mock('react-leaflet-geoman-v2', () => ({
  GeomanControls: (): React.ReactElement => <></>
}))

it('MessageCreator renders correctly', () => {
  const channel = watuWargame.data.channels.channels[0]
  const channels = {}
  watuWargame.data.channels.channels.forEach(c => {
    channels[c.name] = {
      cData: c,
      name: c.name,
      uniqid: c.uniqid
    } as ChannelUI
  })

  const postBack = (details: MessageDetails, message: any): void => {
    console.log('send message', details, message)
  }

  const tree = renderer
    .create(<MessageCreator
      channel={channel}
      messageOption='Chat'
      currentTurn={0}
      gameDate=''
      privateMessage={true}
      schema={MessageTemplatesMock[0].details}
      selectedForce={watuWargame.data.forces.forces[0]}
      selectedRole={watuWargame.data.forces.forces[0].roles[0].roleId}
      selectedRoleName={watuWargame.data.forces.forces[0].roles[0].name}
      postBack={postBack}
    />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
