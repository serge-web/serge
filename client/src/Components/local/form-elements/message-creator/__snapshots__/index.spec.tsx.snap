// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageCreator renders correctly 1`] = `
Array [
  <div
    className="alert alert-danger fade"
    role="alert"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <h5>
      Schema:
    </h5>
     
    
  </div>,
  <div
    className="alert alert-danger fade"
    role="alert"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <h5>
      UI Schema:
    </h5>
     
    
  </div>,
  <form
    className="form-group message-creator edt-disable"
    noValidate={false}
    onSubmit={[Function]}
  >
    <div
      className="form-group field field-object"
    >
      <div
        className="form-group"
      >
        <div
          className="my-1"
          id="root__title"
        >
          <h5>
            Chat
          </h5>
          <hr
            className="border-0 bg-secondary"
            style={
              Object {
                "height": "1px",
              }
            }
          />
        </div>
        <div
          className="p-0 container-fluid"
        >
          <div
            className="row"
            style={
              Object {
                "marginBottom": "10px",
              }
            }
          >
            <div
              className="col-12"
            >
               
              <div
                className="form-group field field-boolean"
              >
                <div
                  className="form-group"
                >
                  <input
                    className="unchecked"
                    onClick={[Function]}
                    required={false}
                    type="checkbox"
                  />
                  <label>
                    completed
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            className="row"
            style={
              Object {
                "marginBottom": "10px",
              }
            }
          >
            <div
              className="col-12"
            >
               
              <div
                className="form-group field field-string"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                    htmlFor="root_Reference"
                  >
                    Reference
                  </label>
                  <input
                    aria-describedby="root_Reference__error root_Reference__description root_Reference__help"
                    autoFocus={false}
                    className="form-control"
                    disabled={false}
                    id="root_Reference"
                    name="root_Reference"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder=""
                    readOnly={false}
                    required={false}
                    type="text"
                    value=""
                  />
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="row"
            style={
              Object {
                "marginBottom": "10px",
              }
            }
          >
            <div
              className="col-12"
            >
               
              <div
                className="form-group field field-string"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                    htmlFor="root_Title"
                  >
                    Title
                  </label>
                  <input
                    aria-describedby="root_Title__error root_Title__description root_Title__help"
                    autoFocus={false}
                    className="form-control"
                    disabled={false}
                    id="root_Title"
                    name="root_Title"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder=""
                    readOnly={false}
                    required={false}
                    type="text"
                    value=""
                  />
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="row"
            style={
              Object {
                "marginBottom": "10px",
              }
            }
          >
            <div
              className="col-12"
            >
               
              <div
                className="form-group field field-array"
              >
                <div
                  className="form-group"
                >
                  <div>
                    <div
                      className="p-0 m-0 row"
                    >
                      <div
                        className="p-0 m-0 col"
                      >
                        <div
                          className="my-1"
                          id="root_Forces__title"
                        >
                          <h5>
                            Forces
                          </h5>
                          <hr
                            className="border-0 bg-secondary"
                            style={
                              Object {
                                "height": "1px",
                              }
                            }
                          />
                        </div>
                        <div
                          className="p-0 m-0 container-fluid"
                        >
                          <div
                            className="container"
                          >
                            <div
                              className="mt-2 row"
                            >
                              <div
                                className="col-9"
                              />
                              <div
                                className="py-4 col-lg-3 col-3 col-3"
                              >
                                <button
                                  className="ml-1 array-item-add btn btn-primary"
                                  disabled={false}
                                  onClick={[Function]}
                                  style={
                                    Object {
                                      "width": "100%",
                                    }
                                  }
                                  title="Add Item"
                                  type="button"
                                >
                                  <svg
                                    fill="currentColor"
                                    height="1em"
                                    stroke="currentColor"
                                    strokeWidth="0"
                                    style={
                                      Object {
                                        "color": undefined,
                                      }
                                    }
                                    viewBox="0 0 16 16"
                                    width="1em"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      clipRule="evenodd"
                                      d="M8 3.5a.5.5 0 01.5.5v4a.5.5 0 01-.5.5H4a.5.5 0 010-1h3.5V4a.5.5 0 01.5-.5z"
                                      fillRule="evenodd"
                                    />
                                    <path
                                      clipRule="evenodd"
                                      d="M7.5 8a.5.5 0 01.5-.5h4a.5.5 0 010 1H8.5V12a.5.5 0 01-1 0V8z"
                                      fillRule="evenodd"
                                    />
                                  </svg>
                                </button>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-action btn-action--form btn-action--cancel"
        name="cancel"
        onClick={[Function]}
        type="button"
      >
        <span>
          Cancel
        </span>
      </button>
      <button
        className="btn btn-action btn-action--form btn-action--send-message"
        name="send"
      >
        <span>
          Send Message
        </span>
      </button>
    </div>
  </form>,
  <div
    className="flex-content form-group"
  >
    <label
      className="material-label"
      htmlFor=""
      id="private-message-input-label"
    >
      <svg
        className="svg-inline--fa fa-user-secret"
      />
      Private message
    </label>
    <textarea
      className="form-control"
      id="private-message-input"
      onChange={[Function]}
      value=""
    />
  </div>,
]
`;
